File: src/lib/validators/endpoint.ts
import { z } from "zod";

// Allow empty strings to pass through as undefined for optional URL/email fields
const optionalEmail = z.union([
  z.string().email(),
  z.literal(""),
]).transform((v) => (v === "" ? undefined : v)).optional();

const optionalUrl = z.union([
  z.string().url(),
  z.literal(""),
]).transform((v) => (v === "" ? undefined : v)).optional();

const endpointFields = {
  projectId: z.string().nullable().optional(),
  name: z.string().min(1).max(100),
  url: z.string().url(),
  method: z.enum(["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"]),
  expectedStatusCode: z.number().int().min(100).max(599),
  interval: z.number().int().min(60).max(3600),
  timeout: z.number().int().min(1).max(60),
  headers: z.record(z.string(), z.string()).optional(),
  body: z.string().optional(),
  isActive: z.boolean().optional(),
  alertEnabled: z.boolean().optional(),
  alertThreshold: z.number().int().min(1).max(100).optional(),
  notifications: z
    .object({
      email: z
        .object({
          enabled: z.boolean(),
          address: optionalEmail,
        })
        .optional(),
      discord: z
        .object({
          enabled: z.boolean(),
          webhookUrl: optionalUrl,
        })
        .optional(),
      webhook: z
        .object({
          enabled: z.boolean(),
          url: optionalUrl,
        })
        .optional(),
    })
    .optional(),
};

export const createEndpointSchema = z.object(endpointFields);

// For updates: all fields optional, ignore extra mongoose fields
export const updateEndpointSchema = z.object(endpointFields).partial();

File: backend/schemas/endpoint.py
from typing import Literal, Optional

from pydantic import BaseModel, Field


class EmailNotificationSchema(BaseModel):
    enabled: bool = False
    address: Optional[str] = None


class DiscordNotificationSchema(BaseModel):
    enabled: bool = False
    webhook_url: Optional[str] = None


class WebhookNotificationSchema(BaseModel):
    enabled: bool = False
    url: Optional[str] = None


class NotificationsSchema(BaseModel):
    email: EmailNotificationSchema = EmailNotificationSchema()
    discord: DiscordNotificationSchema = DiscordNotificationSchema()
    webhook: WebhookNotificationSchema = WebhookNotificationSchema()


class CreateEndpointRequest(BaseModel):
    project_id: Optional[str] = None
    name: str
    url: str
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"] = "GET"
    expected_status_code: int = 200
    interval: int = Field(default=60, ge=60, le=3600)
    timeout: int = Field(default=10, ge=1, le=60)
    headers: dict[str, str] = Field(default_factory=dict)
    body: str = ""
    is_active: bool = True
    alert_enabled: bool = False
    alert_threshold: int = 3
    notifications: NotificationsSchema = NotificationsSchema()


class UpdateEndpointRequest(BaseModel):
    project_id: Optional[str] = None
    name: Optional[str] = None
    url: Optional[str] = None
    method: Optional[Literal["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"]] = None
    expected_status_code: Optional[int] = None
    interval: Optional[int] = Field(None, ge=60, le=3600)
    timeout: Optional[int] = Field(None, ge=1, le=60)
    headers: Optional[dict[str, str]] = None
    body: Optional[str] = None
    is_active: Optional[bool] = None
    alert_enabled: Optional[bool] = None
    alert_threshold: Optional[int] = None
    notifications: Optional[NotificationsSchema] = None
